core/ipc/sink/aws-sqs-impl/src/main/java/org/opennms/core/ipc/sink/aws/sqs/AmazonSQSRemoteMessageDispatcherFactory.java

```diff
@@ -0,0 +1,183 @@
+ *     OpenNMS(R) Licensing <license@opennms.org>
+ *     http://www.opennms.org/
+ *     http://www.opennms.com/
+ *******************************************************************************/
+
+package org.opennms.core.ipc.sink.aws.sqs;
+
+import java.io.IOException;
+import java.util.Dictionary;
+import java.util.Enumeration;
+import java.util.Properties;
+
+import org.opennms.core.ipc.sink.api.Message;
+import org.opennms.core.ipc.sink.api.MessageConsumerManager;
+import org.opennms.core.ipc.sink.api.SinkModule;
+import org.opennms.core.ipc.sink.common.AbstractMessageDispatcherFactory;
+import org.opennms.core.logging.Logging;
+import org.opennms.core.logging.Logging.MDCCloseable;
+import org.osgi.service.cm.ConfigurationAdmin;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.amazonaws.services.sqs.AmazonSQS;
+import com.amazonaws.services.sqs.model.AmazonSQSException;
+import com.amazonaws.services.sqs.model.QueueDoesNotExistException;
+import com.codahale.metrics.JmxReporter;
+
+/**
+ * A factory for creating AwsRemoteMessageDispatcher objects.
+ * 
+ * @author <a href="mailto:agalue@opennms.org">Alejandro Galue</a>
+ */
+public class AmazonSQSRemoteMessageDispatcherFactory extends AbstractMessageDispatcherFactory<String> {
+
+    /** The Constant LOG. */
+    private static final Logger LOG = LoggerFactory.getLogger(AmazonSQSRemoteMessageDispatcherFactory.class);
+
+    /** The AWS configuration. */
+    private final Properties awsConfig = new Properties();
+
+    /** The configuration administration object. */
+    private ConfigurationAdmin configAdmin;
+
+    /** The reporter. */
+    private JmxReporter reporter;
+
+    /** The SQS Object. */
+    private AmazonSQS sqs;
+
+    /* (non-Javadoc)
+     * @see org.opennms.core.ipc.sink.common.AbstractMessageDispatcherFactory#getModuleMetadata(org.opennms.core.ipc.sink.api.SinkModule)
+     */
+    @Override
+    public <S extends Message, T extends Message> String getModuleMetadata(final SinkModule<S, T> module) {
+        return AmazonSQSUtils.getQueueName(module);
+    }
+
+    /* (non-Javadoc)
+     * @see org.opennms.core.ipc.sink.common.AbstractMessageDispatcherFactory#dispatch(org.opennms.core.ipc.sink.api.SinkModule, java.lang.Object, org.opennms.core.ipc.sink.api.Message)
+     */
+    @Override
+    public <S extends Message, T extends Message> void dispatch(SinkModule<S, T> module, String topic, T message) {
+        try (MDCCloseable mdc = Logging.withPrefixCloseable(MessageConsumerManager.LOG_PREFIX)) {
+            LOG.trace("dispatch({}): sending message {}", topic, message);
+            final String queueName = getModuleMetadata(module);
+            final String queueUrl = getQueueUrl(queueName);
+            if (queueUrl == null) {
+                LOG.warn("Cannot obtain URL for queue {}. The message cannot be sent.", queueName);
+            } else {
+                sqs.sendMessage(queueUrl, module.marshal((T)message));
```

Comment:
How does this behave if SQS is not reachable?

Reply:
The API doesn't say anything about it, and there is no sign that the method will thrown an exception, so maybe silently ignored ? I can put a try/catch around it, catching `Exception` maybe ?

Reply:
Generally speaking, the implementation of the dispatch method should block until the message is successfully delivered.

I think we need to validate how the SQS client behaves if for example the Minion were to lose internet connectivity and continue to receive SNMP traps.